### ISSUE TITLE
dnx hangs in docker container

### ISSUE THREAD

**Issue by rohitshrivastava04** [dotnet/coreclr#2466]
I am trying to build docker container with dev build of dotnet coreclr following is my docker file

    FROM debian:jessie

    ENV DNX_VERSION 1.0.0-rc2-16351
    ENV DNX_USER_HOME /opt/DNX_BRANCH
    #Currently the CLR packages don't have runtime ids to handle debian:jessie but
    #we are making sure that the dependencies are the right versions and are opting for
    #the smaller base image. So we use this variable to overwrite the default detection.
    ENV DNX_RUNTIME_ID ubuntu.14.04-x64

    RUN apt-get -qq update && apt-get -qqy install unzip curl libicu-dev libunwind8 gettext libssl-dev libcurl3-gnutls zlib1g && rm -rf /var/lib/apt/lists/*

    RUN curl -sSL https://raw.githubusercontent.com/aspnet/Home/dev/dnvminstall.sh | DNX_USER_HOME=$DNX_USER_HOME DNX_BRANCH=v$DNX_VERSION sh
    RUN bash -c "source $DNX_USER_HOME/dnvm/dnvm.sh \
	&& dnvm install $DNX_VERSION -alias default -r coreclr -u\
	&& dnvm alias default | xargs -i ln -s $DNX_USER_HOME/runtimes/{} $DNX_USER_HOME/runtimes/default"

    # Install libuv for Kestrel from source code (binary is not in wheezy and one in jessie is still too old)
    # Combining this with the uninstall and purge will save us the space of the build tools in the image
    RUN LIBUV_VERSION=1.4.2 \
	&& apt-get -qq update \
	&& apt-get -qqy install autoconf automake build-essential libtool \
	&& curl -sSL https://github.com/libuv/libuv/archive/v${LIBUV_VERSION}.tar.gz | tar zxfv - -C /usr/local/src \
	&& cd /usr/local/src/libuv-$LIBUV_VERSION \
	&& sh autogen.sh && ./configure && make && make install \
	&& rm -rf /usr/local/src/libuv-$LIBUV_VERSION \
	&& ldconfig \
	&& apt-get -y purge autoconf automake build-essential libtool \
	&& apt-get -y autoremove \
	&& apt-get -y clean \
	&& rm -rf /var/lib/apt/lists/*

    ENV PATH $PATH:$DNX_USER_HOME/runtimes/default/bin

Now when I build my app image for docker using the above image dnx command hangs after its output and doesn't complete.

Following is my docker file

    FROM aspnet-rc2

    ADD . /app

    WORKDIR /app/approot

    RUN [ "dnu", "restore", "--runtime", "ubuntu.14.04-x64", "--source", "https://www.myget.org/F/dotnet-core/", "--fallbacksource", "https://www.nuget.org/api/v2/"]

    EXPOSE 7600

    ENTRYPOINT ["./web"]

Following is the image which is the output of dnx --version, it never completes.

![image](https://cloud.githubusercontent.com/assets/3762001/11996166/b11c6f7c-aa55-11e5-9627-0abba606438c.png)

Where as rc1-update1 works fine. If there is open bug around this, could you please tell me last RC2 build which works.

This can easily be reproduced if you create a container using 
    docker run -i -t debian:jessie
and isntall DNVM on it and run dnvm upgrade -r coreclr -u

and execute dnx --version

Please help!

**Comment by janvorli** [comment]
@rohitshrivastava04 Can you check your locale settings please? Just run "locale" in your shell and paste the result here. I am looking into a similar issue (#2215) and in that case, the issue is caused by the locale being `POSIX` or `C` where the resource manager is trying to look up resources in an non-existent resource dll instead of using the default ones. `en-US` locale works fine.

**Comment by rohitshrivastava04** [comment]
yes, you are correct its POSIX.
    locale
    LANG=
    LANGUAGE=
    LC_CTYPE="POSIX"
    LC_NUMERIC="POSIX"
    LC_TIME="POSIX"
    LC_COLLATE="POSIX"
    LC_MONETARY="POSIX"
    LC_MESSAGES="POSIX"
    LC_PAPER="POSIX"
    LC_NAME="POSIX"
    LC_ADDRESS="POSIX"
    LC_TELEPHONE="POSIX"
    LC_MEASUREMENT="POSIX"
    LC_IDENTIFICATION="POSIX"
    LC_ALL=

Could you please tell me how can I change it and what it should be?

**Comment by janvorli** [comment]
It is a bug that it doesn't work with POSIX as well and it needs to be fixed. But you can work around the issue by setting the locale this way:
```sh
export LANG=en_US.UTF-8
export LANGUAGE=en_US.UTF-8
export LC_ALL=en_US.UTF-8
```

**Comment by rohitshrivastava04** [comment]
Thanks for your reply. I tried changing locale however even after changing locale dnx --version command hangs.
Could you please provide me a docker file if possible?

**Comment by janvorli** [comment]
Ok, maybe it is some other issue then. As for the docker file, my knowledge of docker is next to none. When looking into the #2215, I have just used a docker package that @Sridhar-MS has created for me, used `docker run` command to run it, then set the locale and launched dnx restore. 
So I am unable to provide you with a docker file at the moment.

**Comment by rohitshrivastava04** [comment]
@Sridhar-MS Could you please pass me the docker file which you used to build CoreCLR docker image?

**Comment by rohitshrivastava04** [comment]
Is there a way to see whats its looking for?

**Comment by jasonwilliams200OK** [comment]
@rohitshrivastava04, AFAIK, most of the Docker stuff and the information you are looking for is available at https://github.com/dotnet/cli. 
@davidfowl, is asp.net Docker part of dotnet-cli repo too?

**Comment by rohitshrivastava04** [comment]
How can I install dotnet cli RC2? I checked and found that if I install dotnet cli I get  version CoreClr-x64-1.0.0-rc1-16048 where as I want to use RC2.

**Comment by janvorli** [comment]
@rohitshrivastava04 I have a PR for a real fix for the  #2215 (PR #2528) - the locale turned out to be a red herring in that case. I believe it is highly probable that it will fix your problem as well. 
So I'd recommend you to wait for the fix.

**Comment by janvorli** [comment]
@rohitshrivastava04 the fix is in

**Comment by rohitshrivastava04** [comment]
Thanks @janvorli 

Do I have to wait for nightly build to get dnvm upgrade -r coreclr -u to get these changes? Could you please tell me what would be the coreclr version which will have these changes?

**Comment by rohitshrivastava04** [comment]
I managed to change the unstable feed url to https://www.myget.org/F/aspnetcidev/api/v2, but still dnx hangs in 1.0.0-rc2-16375

**Comment by janvorli** [comment]
@rohitshrivastava04 I would like to try to repro it locally - can you please provide me with step by step instructions? I've ran `sudo docker run -i -t debian:jessie` as you've said, but I don't know how to install dnvm on it. Attempt to use instructions from http://docs.asp.net/en/latest/getting-started/installing-on-linux.html fails since `apt-get install unzip curl` fails to locate the unzip and curl.

**Comment by rohitshrivastava04** [comment]
First update apt-get by running following command

    apt-get -qq update 

Then run 

    apt-get -qqy install unzip curl libicu-dev libunwind8 gettext libssl-dev libcurl3-gnutls zlib1g && rm -rf /var/lib/apt/lists/*
    curl -sSL https://raw.githubusercontent.com/aspnet/Home/dev/dnvminstall.sh | DNX_BRANCH=dev sh
    source ~/.dnx/dnvm/dnvm.sh
    dnvm upgrade -r coreclr -u
   

And then when your run dnvm list it will show that rc2 is installed, now if you run 

    dnx --version

it will hang.

To install dnx from dev branch you can change the environment variable

    export DNX_UNSTABLE_FEED=https://www.myget.org/F/aspnetcidev/api/v2

Install latest rc2  

    dnvm upgrade -r coreclr -u

**Comment by janvorli** [comment]
@rohitshrivastava04, thanks for the instructions. I've tried the steps and the `dnx --version` didn't hang for me:
```
root@da7ed9815554:/# dnx --version
Microsoft .NET Execution environment
 Version:      1.0.0-rc2-16379
 Type:         CoreClr
 Architecture: x64
 OS Name:      debian
 OS Version:   8
 Runtime Id:   debian.8-x64

root@da7ed9815554:/# 
```
So hopefuly the fix got finally propagated.

**Comment by rohitshrivastava04** [comment]
I am using docker on windows 7, could you please confirm your environment as mine still hangs.

**Comment by janvorli** [comment]
@rohitshrivastava04 I am not sure I understand - we were discussing linux stuff here, so I am not sure how Windows 7 is related.

**Comment by rohitshrivastava04** [comment]
I am using boot2docker on windows 7.

-Rohit

**Comment by rohitshrivastava04** [comment]
@janvorli thanks for pointing me in the right direction. Now I can confirm that its an issue only when using boot2docker on Windows 7. I installing on ubuntu as VM on my windows 7 and installed docker on it and it ran smooth. First time I can see something which is coming from Microsoft but throws an error on Windows and runs fine on Ubuntu. I am glad to see this that primary focus of CoreCLR is Linux.

**Comment by sergiy-k** [comment]
Hi @rohitshrivastava04, are you still seeing this issue with dnx on Windows7?

**Comment by rohitshrivastava04** [comment]
Yes, I tested it today again on windows 7 with boot2docker and it still hangs.

**Comment by janvorli** [comment]
@sergiy-k Please note this is not a supported Linux distro. The boot2docker is some very minimalistic Linux distro.

**Comment by billycc** [comment]
@rohitshrivastava04 @janvorli  i'm having a similar issue (trying to build a docker based off of aspnet/1.0.0-rc1-update1-coreclr and restoring sqlclient among other things, and it leads to dnx hang). 

i'm using docker 1.9.1 (no boot2docker) on win10. i also tried building a similar dockerfile hosting from an ubuntu box.



so, i put together a very hacky script to get over the dnx hang. 

    #!/usr/bin/env bash
    
    #very important#
    #cd into the path to the root of your source tree, like: 
    #cd /app/

    dnu restore --parallel --runtime ubuntu.14.04-x64 2>&1 | tee /tmp/dnu.restore  &
    
    # wait up to 5 minutes (150*2seconds)
    for i in `seq 1 150|xargs`
    do
      sleep 2
      echo -n .
    
      if grep -e '^Error' -e 'Restore failed' /tmp/dnu.restore ;
      then
        exit -1
      fi

      if grep -e 'Total time' /tmp/dnu.restore ;
      then
        sleep 5
        exit 0
      fi
    done

    exit -1

you'll notice dots interspersed with the normal dnu restore output, and eventually (hopefully) the script will exit itself. (and bash should kill the backgrounded dnu-restore if it was hanging)

with that script inside an appropriate place then replace a normal `RUN dnu restore...` with `RUN chmod a+x ./hack-restore.hs && ./hack-restore.sh` or similar.

**Comment by rohitshrivastava04** [comment]
@billycc rc1 + sqlclient does not work. You have to upgrade to RC2. I would recommend you to build rc2 image yourself and then build an application image from that. One more thing, try on ubuntu as on windows dnx hangs. If you want I can give you dockerfile to build aspnet rc2 image.

**Comment by billycc** [comment]
@rohitshrivastava04 yeah. i'm working my way on building up a rc2 container. if you got it all working, then please share. i just mentioned the rc1, because from the looks of it, you had started from aspnet's udpate1-coreclr; which is where i had started in trying to build mine.

my docker begins `FROM microsoft/aspnet:1.0.0-rc1-update1-coreclr`, but then proceeds to do other things including `dnvm install 1.0.0-rc2-...` (still trying a few versions off the diff feeds) and i restore with `hack-restore.sh` to avoid the dnx-hang.

**Comment by billycc** [comment]
on further investigation, that hack script i made doesn't seem to work. fooled myself into thinking it was working, but i was restoring as rc1 earlier on in my dockerfile and that lead to a mixed packages folder.

i was doing restore as rc1 (which i overlooked), and then a later one was restoring as rc2. and the rc2 restore wasn't really hanging in that case. (or, that hack does work in that case.) if i cut out the rc1 restore, then my hack script doesn't kill the hanging dnx. :/

**Comment by jasonwilliams200OK** [comment]
@rohitshrivastava04, it seems like boot2docker is deprecated in favor of Docker Machine (https://docs.docker.com/engine/installation/windows/):

> 
**Note:** This release of Docker deprecates the Boot2Docker command line in favor of Docker Machine. Use the Docker Toolbox to install Docker Machine as well as the other Docker tools.

I have installed docker machine on Windows. Can you share Dockerfile (via https://gist.github.com/) with rc2 stuff so I can try to reproduce the `dnx --version` hanging issue?

**Comment by rohitshrivastava04** [comment]
@jasonwilliams200OK sorry my bad, I am using docker-machine on windows 7. I mentioned boot2docker as docker-machine creates a VM using boot2docker.iso.

https://gist.github.com/rohitshrivastava04/69df942663fce84f62f2

**Comment by tmds** [comment]
Might be relevant here. I recently had some issues using docker with coreclr where processes were hanging and showed up as '<defunct>'. This turned out to be a bug in docker which was fixed in v1.10.

**Comment by sergiy-k** [comment]
It seems there are various topics have been discussed in this issue. May I please ask you to open a separate issue on https://github.com/dotnet/cli/ in case if you are still having problems with installing dotnet/dnx or restoring packages. If no one objects, I would like to close this issue because the problem that was reported in the original post has post has been fixed.
